charts/loki-simple-scalable/templates/write/statefulset-write.yaml

```diff
@@ -14,6 +14,9 @@ spec:
       partition: 0
   serviceName: {{ include "loki.writeFullname" . }}-headless
   revisionHistoryLimit: {{ .Values.loki.revisionHistoryLimit }}
+  persistentVolumeClaimRetentionPolicy:
+    whenDeleted: Retain
+    whenScaled: Delete
```

Comment:
Can you please elaborate (or even better - add inline comments) about why we want to retain a PVC when deleting a `write` pod, but delete it when scaling down?

Reply:
The reason for this is based on my understanding of how the [rollout operator](https://github.com/grafana/rollout-operator) will work. If I understand correctly, since the update strategy is set to `OnDelete`, the stateful sets will not update on their own, and instead the rollout operator will delete the old replica for a given zone and replace it with a new one. In this case, we want to `Retain` the existing PVCs so the new pods can reuse them. When scaling, we want to the `Delete` the PVCs however, as our rings will re-shard any data they were holding to other instances, and that data is no long valid. Setting retention to `Delete` for scaling events will hopefully make those scaling events easier from an operations perspective.

Reply:
Also, this was the best resource I could find on the difference between the different policy types: https://kubernetes.io/blog/2021/12/16/kubernetes-1-23-statefulset-pvc-auto-deletion/#the-new-statefulset-pvc-retention-policy

Reply:
Few comments:
- We're **not** used to set the `persistentVolumeClaimRetentionPolicy` in Mimir
- I think `persistentVolumeClaimRetentionPolicy` also requires K8S 1.23 or above ([doc](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#persistentvolumeclaim-retention))
- According to the doc, `whenDeleted` "configures the volume retention behavior that applies when the StatefulSet is deleted". It's not when the pod is deleted, but when the StatefulSet is deleted. I would be suprised if a PVC would get deleted as soon as a pod is deleted, because the pod deletion is just a normal operation which occurs during rollout even when using the `RollingUpdate` update strategy (the difference is who deletes pod, either the StatefulSet controller or an external entity, like the rollout-operator).

All in all, I think that:
- You don't want to set `persistentVolumeClaimRetentionPolicy` if you want to support K8S < 1.23
- If you set it, you may want to set `Retain` for both `whenDeleted` and `whenScaled` for the **write** path, because if you scale down or delete the StatefulSet by mistake, you don't want to loose WALs (that's the reasoning we did for Mimir ingesters too)